torch/csrc/profiler/README.md - stubs, RecordFunction, Autograd interaction

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108470

Technical details about the profiler - stubs for the stuff I haven't had time to fill out yet, plus details about RecordFunction and the profiler's interaction with autograd.

reviewers - see https://github.com/pytorch/pytorch/blob/06c41eea9e37b5799e4f14601c95fa7810297759/torch/csrc/profiler/README.md for rendered markdown

davidberard98 committed:
[WIP] torch/csrc/profiler/README.md

Technical details about the profiler.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] torch/csrc/profiler/README.md"


Technical details about the profiler.

[ghstack-poisoned]

anupambhatnagar commented:
@davidberard98 Thanks for putting this readme together. Is this PR currently a WIP or are you planning to commit it as is?

davidberard98 commented:
@anupambhatnagar I was planning to land it as is - not sure I have a good enough understanding of the other parts to write about them clearly right now.

aaronenyeshi commented on torch/csrc/profiler/README.md:
```diff
@@ -0,0 +1,56 @@
+The profiler instruments PyTorch to collect information about the model's execution. Its main features are:
+* Instrumenting op calls on the CPU side
+* Interfacing with [Kineto](https://github.com/pytorch/kineto/) to collect information from the GPU (or other accelerators)
+* Collecting python stacktraces
```
```suggestion
* Collecting python stack traces
```

aaronenyeshi commented on torch/csrc/profiler/README.md:
```diff
@@ -0,0 +1,56 @@
+
+RecordFunction is used by the profiler to instrument CPU-side events.
+
+RecordFunction is a general method of instrumenting function calls in pytorch. It can be used for other general applications, e.g. see [Features for Large-Scale Deployments](https://pytorch.org/docs/stable/notes/large_scale_deployments.html). In PyTorch, it is already included at some important locations; notably, in the [dispatcher](https://github.com/pytorch/pytorch/blob/247c603da9b780534e25fb1 ...
```
```suggestion
RecordFunction is a general method of instrumenting function calls in PyTorch. It can be used for other general applications, e.g. see [Features for Large-Scale Deployments](https://pytorch.org/docs/stable/notes/large_scale_deployments.html). In PyTorch, it is already included at some important locations; notably, in the [dispatcher](https://github.com/pytorch/pytorch/blob/247c603da9b780534e25fb1d90b6e5a528b625b1/aten/src/ATen/core/dispatch/Dispatcher.h#L650), surrounding every op.
```

aaronenyeshi commented on torch/csrc/profiler/README.md:
```diff
@@ -0,0 +1,56 @@
+
+RecordFunction is a general method of instrumenting function calls in pytorch. It can be used for other general applications, e.g. see [Features for Large-Scale Deployments](https://pytorch.org/docs/stable/notes/large_scale_deployments.html). In PyTorch, it is already included at some important locations; notably, in the [dispatcher](https://github.com/pytorch/pytorch/blob/247c603da9b780534e25fb1 ...
+
+Users (or PyTorch itself) can register callbacks that will be executed whenever a RecordFunction guard is encountered. The profiler uses this mechanism to record the start and end times for each op call, as well as user-provided RecordFunction anntations. The RecordFunction machinery is designed to have relatively low overhead, especially when there are no callbacks registered. Nevertheless, ther ...
```
```suggestion
Users (or PyTorch itself) can register callbacks that will be executed whenever a RecordFunction guard is encountered. The profiler uses this mechanism to record the start and end times for each op call, as well as user-provided RecordFunction annotations. The RecordFunction machinery is designed to have relatively low overhead, especially when there are no callbacks registered. Nevertheless, there can still be some overhead.
```

aaronenyeshi reviewed (approved):
Overall looks great, added a few small comments. This is a great addition to the profiler docs, and a good start for further sections.

Also, wondering since RecordFunction is a C++ struct, should we annotate it as code? Like `RecordFunction` ?

davidberard98 committed:
Update on "torch/csrc/profiler/README.md - stubs, RecordFunction, Autograd interaction"


Technical details about the profiler - stubs for the stuff I haven't had time to fill out yet, plus details about RecordFunction and the profiler's interaction with autograd.

reviewers - see https://github.com/pytorch/pytorch/blob/06c41eea9e37b5799e4f14601c95fa7810297759/torch/csrc/profiler/README.md for rendered markdown

[ghstack-poisoned]

davidberard98 committed:
Update on "torch/csrc/profiler/README.md - stubs, RecordFunction, Autograd interaction"


Technical details about the profiler - stubs for the stuff I haven't had time to fill out yet, plus details about RecordFunction and the profiler's interaction with autograd.

reviewers - see https://github.com/pytorch/pytorch/blob/06c41eea9e37b5799e4f14601c95fa7810297759/torch/csrc/profiler/README.md for rendered markdown

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
